src/main/haskell/kore/src/Kore/Reflect.hs

```diff
@@ -0,0 +1,351 @@
+                 ( Const )
+import           Data.Functor.Foldable
+                 ( Fix (..) )
+import           Data.Functor.Identity
+                 ( Identity (Identity) )
+import           Data.Hashable
+                 ( Hashable )
+import           Data.List
+                 ( intercalate )
+import qualified Data.Map as Map
+import           Data.Sequence
+                 ( Seq )
+import qualified Data.Set as Set
+import           Data.Text
+                 ( Text )
+import           Data.Void
+                 ( Void )
+import           GHC.Generics
+                 ( Constructor, Datatype, Fixity (..), Generic, Selector )
+import qualified GHC.Generics as Generics
+import           Numeric.Natural
+                 ( Natural )
+import           Text.Show.Deriving
+                 ( deriveShow1 )
+
+{-| An reflected element that has a name and a single value, e.g. a struct
+field value.
+-}
+data NamedElement v = NamedElement {name :: !String, value :: !v}
+    deriving (Show, Eq, Generic)
+
+deriveEq1 ''NamedElement
+deriveShow1 ''NamedElement
+
+instance Hashable v => Hashable (NamedElement v)
+
+{-| A reflected element that has a name and any number of values,
+including none, e.g. product types.
+-}
+data ProductElement v = ProductElement {name :: !String, values :: ![v]}
+    deriving (Show, Eq, Generic)
+
+deriveEq1 ''ProductElement
+deriveShow1 ''ProductElement
+
+instance Hashable v => Hashable (ProductElement v)
+
+{-| One level of a reflected value, with v as the value's children.
+
+As an example, when reflecting
+
+@
+SomeStruct {someField1 = value1, someField1 = value2}
+@
+
+then data would contain the representation of
+
+@
+SomeStruct {someField1 = ?, someField1 = ?}
+@
+
+with the value of question marks depending on the v parameter.
+-}
+data Data v
+    = Sum !(ProductElement v)
+    -- ^ Represents a value of the form (Name childValue1 ... childValue_n)
+    | Struct !String ![NamedElement v]
+    -- ^ Represents struct value.
+    | Value !String
+    -- ^ Represents a leaf value, i.e. one without children
+    | List ![v]
+    -- ^ Represents a list of values
+    | Tuple ![v]
+    -- ^ Represents a tuple of values
+    deriving (Show, Eq, Generic)
+
+deriveEq1 ''Data
+deriveShow1 ''Data
+
+instance Hashable v => Hashable (Data v)
+
+{-| A reflected value.
+-}
+type RecursiveData = Fix Data
+
+{-| Class for things that can be represented as 'RecursiveData'.
+
+For any instance of Generic, if the data's "children" (see below) are
+Reflectable, the 'reflect' fuction is automatically implemented and the
+derivation looks something like:
+
+@
+instance
+    (constraints-to-make-MyData-generic)
+    => (Reflectable MyData)
+@
+
+A data type's children are, roughly, the parts that are not declared in
+that data's declaration. Examples:
+
+@
+data MyData = MyData T1 T2
+-- Children: T1 and T2
+
+data MyData = MyData {f1 :: T1, f2 :: T2}
+-- Children: T1 and T2
+@
+-}
+class Reflectable a where
```

Comment:
Does this need to be a new class, since it can be implemented only in terms of `Generic`?

Reply:
I am implementing it more or less directly for Id and Pattern and the basic types. I'll probably drop the Id instance and I might actually manage to make it work for Pattern (I'm not sure why I failed before, but it looked like it's hard to make Generic play nice with GADTs), but I'm not sure it would be better to implement it through Generic for, say, lists, sets, tuples and other similar things (my guess is that each of these would need a chunk of long custom code).